### PR TITLE
Security

### LINKED ISSUE #14: Create JWT
Should create and send a JWT token with the user request

### PR DESCRIPTION
**Pull request by Xander9**
This module contains necessary configurations for user authentication. Requires merging with user, industry and product management for testing and to continue development.

Closes #14

### CODE REVIEW

File: Backend/sprintone/src/main/java/com/project/omega/bean/dao/JwtResponse.java [hunk 1]
```diff
@@ -0,0 +1,17 @@
+package com.project.omega.bean.dao;
+
+import java.io.Serializable;
+
+public class JwtResponse implements Serializable {
+
+    private static final long serialVersionUID = -8091879091924046844L;
+    private final String jwtToken;
+
```
Comment (travelingdeveloper12) [inline Backend/sprintone/src/main/java/com/project/omega/bean/dao/JwtResponse.java:9]: JwtResponse class should also return a Date variable which signifies the date and time of the generated token. The returned value should always be enclosed in the toString() method.

File: Backend/sprintone/src/main/java/com/project/omega/config/JwtAuthenticationEntryPoint.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+@Component
+public class JwtAuthenticationEntryPoint implements AuthenticationEntryPoint, Serializable {
+
+    private static final long serialVersionUID = -7858869558953243875L;
+
+    @Override
+    public void commence(HttpServletRequest request, HttpServletResponse response,
+                         AuthenticationException authException) throws IOException {
+
+        response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Unauthorized");
```
Comment (travelingdeveloper12) [inline Backend/sprintone/src/main/java/com/project/omega/config/JwtAuthenticationEntryPoint.java:21]: Hardcoding should be avoided any description within double quotes should be referenced explicitly either through a Constants class or should be referenced as a static import

File: Backend/sprintone/src/main/java/com/project/omega/config/WebSecurityConfig.java [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+    public AuthenticationManager authenticationManagerBean() throws Exception {
+        return super.authenticationManagerBean();
+    }
+
+    @Override
+    protected void configure(HttpSecurity httpSecurity) throws Exception {
+        httpSecurity.csrf().disable()
+                .authorizeRequests()
+                .antMatchers("/authenticate", "/register").permitAll()
+                .anyRequest().authenticated()
```
Comment (travelingdeveloper12) [inline Backend/sprintone/src/main/java/com/project/omega/config/WebSecurityConfig.java:53]: PreAuthorize() with roles have to implemented and to be saved in a different table.

**Comment by markscamilleri** [comment]
I don't think it makes sense to have the code in a `sprintone` folder

File: Backend/sprintone/src/main/java/com/project/omega/bean/dao/JwtResponse.java (hunk 1 above)
Comment (markscamilleri) [reply Backend/sprintone/src/main/java/com/project/omega/bean/dao/JwtResponse.java:9]: Why would this be necessary? Expiry date is more useful, a day I think? @HarryPi can you confirm?

File: Backend/sprintone/src/main/java/com/project/omega/config/JwtRequestFilter.java [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+            jwtToken = requestTokenHeader.substring(7);
+            try {
+                username = jwtTokenUtil.getUsernameFromToken(jwtToken);
+            } catch (IllegalArgumentException e) {
+                System.out.println("Unable to get JWT Token");
+            } catch (ExpiredJwtException e) {
+                System.out.println("JWT Token has expired");
+            }
+        } else {
+            logger.warn("JWT Token does not begin with Bearer String");
```
Comment (markscamilleri) [inline Backend/sprintone/src/main/java/com/project/omega/config/JwtRequestFilter.java:45]: Better not output to System.out (logger is favoured here). But also `logger` doesn't seem to be defined

File: Backend/sprintone/src/test/java/com/project/OmegaApplicationTests.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@SpringBootTest
+@RunWith(SpringRunner.class)
+public class OmegaApplicationTests {
+
+    @Test
+    public void contextLoads() {
+    }
```
Comment (markscamilleri) [inline Backend/sprintone/src/test/java/com/project/OmegaApplicationTests.java:15]: Where are your tests? So many classes, and none of them are tested...

File: Backend/sprintone/src/main/java/com/project/omega/controller/DummyController.java [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+import org.springframework.stereotype.Repository;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.ArrayList;
+import java.util.List;
+
+@RestController
+public class DummyController {
```
Comment (markscamilleri) [inline Backend/sprintone/src/main/java/com/project/omega/controller/DummyController.java:12]: What is the purpose of this?

File: Backend/sprintone/src/main/java/com/project/omega/controller/JwtAuthenticationController.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+                .loadUserByUsername(authenticationRequest.getEmail());
+
+        final String token = jwtTokenUtil.generateToken(userDetails);
+
+        return ResponseEntity.ok(new JwtResponse(token));
+    }
+
+    @PostMapping(value = "/register", headers = "Accept=application/json")
+    public ResponseEntity createUser(@RequestBody UserDTO user) {
+        return new ResponseEntity(HttpStatus.NOT_IMPLEMENTED);
```
Comment (markscamilleri) [inline Backend/sprintone/src/main/java/com/project/omega/controller/JwtAuthenticationController.java:49]: Let's not forget this as we need it.

File: Backend/sprintone/src/main/java/com/project/omega/enums/RoleEnum.java [hunk 8]
```diff
@@ -0,0 +1,5 @@
+package com.project.omega.enums;
+
+public enum RoleEnum {
```
Comment (markscamilleri) [inline Backend/sprintone/src/main/java/com/project/omega/enums/RoleEnum.java:3]: Roles need to be stored in the database as each of our clients will have the ability to create custom roles. The only roles in there that cannot be changed are `Admins` (which are the admins of our clients), and `Super Admins` (which are admins in our team)

File: Backend/sprintone/src/main/java/com/project/omega/service/JwtUserDetailsService.java [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+
+@Service
+public class JwtUserDetailsService implements UserDetailsService {
+
+    @Autowired
+    private PasswordEncoder bCryptEncoder;
+
+    @Override
+    public UserDetails loadUserByUsername(String email) throws UsernameNotFoundException {
+        if ("a@a.com".equals(email)) {
```
Comment (markscamilleri) [inline Backend/sprintone/src/main/java/com/project/omega/service/JwtUserDetailsService.java:21]: What's this?

Testing code should go in a test suite, where the test suite itself can define a test database with test data

File: Backend/sprintone/src/main/resources/application.properties [hunk 10]
```diff
@@ -0,0 +1,17 @@
+## Port
+server.port = 5000
+endpoints.env.enabled=true
+
+## MySQL
+spring.datasource.url = jdbc:mysql://localhost:3306/esrs_test
+spring.datasource.username = root
+spring.datasource.password = password
```
Comment (markscamilleri) [inline Backend/sprintone/src/main/resources/application.properties:8]: NO!

Many times I've seen this. Many times I've seen databases with default credentials get purged by someone with a script trying random databases for ransom... We need a secure password

Also, do not push these credentials in git at all. Save them as encrypted variables in the Azure Pipeline to replace **only on deploy**. (https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables?view=azure-devops&tabs=yaml%2Cbatch)

Remember, this git repo is public

File: Backend/sprintone/src/main/java/com/project/omega/bean/dao/JwtResponse.java (hunk 1 above)
Comment (HarryPi) [reply Backend/sprintone/src/main/java/com/project/omega/bean/dao/JwtResponse.java:9]: Yes we agreed the expiration duration would be 24 hours.

File: Backend/sprintone/src/main/java/com/project/omega/config/JwtRequestFilter.java (hunk 4 above)
Comment (travelingdeveloper12) [reply Backend/sprintone/src/main/java/com/project/omega/config/JwtRequestFilter.java:45]: The serialversionUID has nothing to do with the Expiry date - if that's what you're referring to. It ensures that during deserialization the same class is loaded.

File: Backend/sprintone/src/test/java/com/project/OmegaApplicationTests.java (hunk 5 above)
Comment (travelingdeveloper12) [reply Backend/sprintone/src/test/java/com/project/OmegaApplicationTests.java:15]: The test folder within the project should have all the test cases. The Contextloads() is meant to work as runners. Any listeners, when included, should have contextloads.

File: Backend/sprintone/src/main/java/com/project/omega/controller/DummyController.java (hunk 6 above)
Comment (travelingdeveloper12) [reply Backend/sprintone/src/main/java/com/project/omega/controller/DummyController.java:12]: This is can be removed. This was just to check if the controller is reachable or not.,

File: Backend/sprintone/src/main/java/com/project/omega/enums/RoleEnum.java (hunk 8 above)
Comment (travelingdeveloper12) [reply Backend/sprintone/src/main/java/com/project/omega/enums/RoleEnum.java:3]: This file should be there, follow the code I committed last night. I have asked Alex to close the existing pull request and follow one true branch.

File: Backend/sprintone/src/main/java/com/project/omega/service/JwtUserDetailsService.java (hunk 9 above)
Comment (travelingdeveloper12) [reply Backend/sprintone/src/main/java/com/project/omega/service/JwtUserDetailsService.java:21]: This file was removed

File: Backend/sprintone/src/main/resources/application.properties (hunk 10 above)
Comment (travelingdeveloper12) [reply Backend/sprintone/src/main/resources/application.properties:8]: @Xander9 - This shouldn't be here.

**Pull request closed without merging**